frame/scheduler/src/lib.rs

```diff
@@ -0,0 +1,487 @@
+}
+
+decl_storage! {
+	trait Store for Module<T: Trait> as Scheduler {
+		/// Any period items.
+		Periodic: map hasher(twox_64_concat) PeriodicIndex => Option<(T::BlockNumber, u32)>;
+
+		/// Items to be executed, indexed by the block number that they should be executed on.
+		/// This is ordered by `priority`.
+		Agenda: map hasher(twox_64_concat) T::BlockNumber => Vec<Scheduled<<T as Trait>::Call>>;
+
+		/// The next free periodic index.
+		LastPeriodicIndex: PeriodicIndex;
+	}
+}
+
+decl_event!(
+	pub enum Event<T> where <T as system::Trait>::BlockNumber {
+		Scheduled(BlockNumber),
+	}
+);
+
+decl_module! {
+	// Simple declaration of the `Module` type. Lets the macro know what its working on.
+	pub struct Module<T: Trait> for enum Call where origin: <T as system::Trait>::Origin {
+		fn deposit_event() = default;
+
+		fn on_initialize(now: T::BlockNumber) -> Weight {
+			let limit = T::MaximumWeight::get();
+			let mut queued = Agenda::<T>::take(now);
+			queued.sort_by_key(|i| i.priority);
+			let mut result = 0;
+			let unused_items = queued.into_iter()
+				.scan(0, |cumulative_weight, i| {
+					*cumulative_weight += i.call.get_dispatch_info().weight;
+					Some((*cumulative_weight, i))
+				})
+				.filter_map(|(cumulative_weight, i)| {
+					if i.priority <= HARD_DEADLINE || cumulative_weight <= limit {
+						let maybe_reschedule = i.maybe_periodic.map(|p| (p, i.clone()));
+						let _ = i.call.dispatch(system::RawOrigin::Root.into());
+						if let Some((periodic_index, i)) = maybe_reschedule {
+							// Register the next instance of this task
+							if let Some((period, count)) = Periodic::<T>::get(periodic_index) {
+								if count > 1 {
+									Periodic::<T>::insert(periodic_index, (period, count - 1));
+								} else {
+									Periodic::<T>::remove(periodic_index);
+								}
+								Agenda::<T>::append_or_insert(now + period, &[i][..]);
+							}
+						}
+						result = cumulative_weight;
+						None
+					} else {
+						Some(i)
+					}
+				})
+				.collect::<Vec<_>>();
+			let next = now + One::one();
+			if !unused_items.is_empty() {
+				Agenda::<T>::append_or_insert(next, &unused_items[..]);
+			}
+			result
+		}
+	}
+}
+
+/// A type that can be used as a scheduler.
+pub trait Schedule<BlockNumber, Call> {
+	/// Schedule a one-off dispatch to happen at the beginning of some block in the future.
+	fn schedule(when: BlockNumber, priority: Priority, call: Call);
+
+	/// Schedule a dispatch to happen periodically into the future for come number of times.
+	///
+	/// An index is returned so that it can be cancelled at some point in the future.
+	fn schedule_periodic(
+		when: BlockNumber,
+		period: BlockNumber,
+		count: u32,
+		priority: Priority,
+		call: Call,
+	) -> PeriodicIndex;
+
+	/// Cancel a periodic dispatch. If called during its dispatch, it will not be dispatched
+	/// any further. If called between dispatches, the next dispatch will happen as scheduled but
+	/// then it will cease.
```

Comment:
```suggestion
	/// Cancel a periodic dispatch. 
	///
	/// If called during its dispatch, it will not be dispatched
	/// any further. If called between dispatches, the next dispatch will happen as scheduled but
	/// then it will cease.
```